apps/backend/src/lib/external-db-sync.ts

```diff
@@ -0,0 +1,203 @@
+      `Cannot sync table "${tableName}": No upsertQuery configured.`
+    );
+  }
+
+  if (newRows.length === 0) return;
+  const placeholderMatches = upsertQuery.match(/\$\d+/g) ?? [];
+  const expectedParamCount =
+    placeholderMatches.length === 0
+      ? 0
+      : Math.max(...placeholderMatches.map((m) => Number(m.slice(1))));
+
+  if (expectedParamCount === 0) {
+    throw new Error(
+      `upsertQuery for table "${tableName}" contains no positional parameters ($1, $2, ...).` +
+        ` Your mapping must use parameterized SQL.`
+    );
+  }
+  const sampleRow = newRows[0];
+  const { tenancyId: _ignore, ...restSample } = sampleRow;
+  const orderedKeys = Object.keys(restSample);
+
+  if (orderedKeys.length !== expectedParamCount) {
+    throw new Error(
+      ` Column count mismatch for table "${tableName}".\n` +
+        `→ upsertQuery expects ${expectedParamCount} parameters.\n` +
+        `→ internalDbFetchQuery returned ${orderedKeys.length} columns (excluding tenancyId).\n` +
+        `Fix your SELECT column order or your SQL parameter order.`
+    );
+  }
+
+  for (const row of newRows) {
+    const { tenancyId, ...rest } = row;
+    const rowKeys = Object.keys(rest);
+
+    const validShape =
+      rowKeys.length === orderedKeys.length &&
+      rowKeys.every((k, i) => k === orderedKeys[i]);
+
+    if (!validShape) {
+      throw new Error(
+        `  Row shape mismatch for table "${tableName}".\n` +
+          `Expected column order: [${orderedKeys.join(", ")}]\n` +
+          `Received column order: [${rowKeys.join(", ")}]\n` +
+          `Your SELECT must be explicit, ordered, and NEVER use SELECT *.\n` +
+          `Fix the SELECT in internalDbFetchQuery immediately.`
+      );
+    }
+  }
+  for (const row of newRows) {
+    const { tenancyId, ...rest } = row;
+    await externalClient.query(upsertQuery, Object.values(rest));
+  }
+}
+
+
+async function syncMapping(
+  externalClient: Client,
+  mappingId: string,
+  mapping: CompleteConfig["dbSync"]["externalDatabases"][string]["mappings"][string],
+  internalPrisma: PrismaClientTransaction,
+  dbId: string,
+  tenancyId: string,
+) {
+
+  const rawSourceTables: any = (mapping as any).sourceTables;
+  const sourceTables: string[] = rawSourceTables
+    ? Object.values(rawSourceTables)
+    : [];
+
+  const rawTargetPk: any = (mapping as any).targetTablePrimaryKey;
+  const targetTablePrimaryKey: string[] = rawTargetPk
+    ? Object.values(rawTargetPk)
+    : [];
+
+  if (sourceTables.length === 0) {
+    console.error(
+      ` Invalid configuration for mapping #${mappingId}: 'sourceTables' resolved to an empty list.`,
+    );
+    return;
+  }
+
+  if (targetTablePrimaryKey.length === 0) {
+    console.error(
+      ` Invalid configuration for mapping #${mappingId}: 'targetTablePrimaryKey' resolved to an empty list.`,
+    );
+    return;
+  }
+
+  const fetchQuery = mapping.internalDbFetchQuery;
+  if (!fetchQuery || !mapping.targetTable) {
+    return;
+  }
```

Comment:
generally whenever i see something like this it's a bit of an antipattern. why exactly would either of these be undefined, and why does that mean we should just skip a table? a comment can help explain